src/serializer/ResidualHeapVisitor.js

```diff
@@ -504,21 +513,45 @@ export class ResidualHeapVisitor {
     }
   }
 
-  // Visits a binding, if createBinding is true, will always return a ResidualFunctionBinding
-  // otherwise visits + returns the binding only if one already exists.
-  visitBinding(val: FunctionValue, name: string, createBinding: boolean = true): ResidualFunctionBinding | void {
-    let residualFunctionBinding;
+  resolveBinding(val: FunctionValue, name: string): EnvironmentRecord {
     let doesNotMatter = true;
     let reference = this.logger.tryQuery(
       () => Environment.ResolveBinding(this.realm, name, doesNotMatter, val.$Environment),
       undefined
     );
-    let getFromMap = createBinding ? getOrDefault : (map, key, defaultFn) => map.get(key);
     if (
       reference === undefined ||
       Environment.IsUnresolvableReference(this.realm, reference) ||
-      reference.base instanceof GlobalEnvironmentRecord
+      reference.base === this.globalEnvironmentRecord ||
+      reference.base === this.globalEnvironmentRecord.$DeclarativeRecord
     ) {
+      return this.globalEnvironmentRecord;
+    } else {
+      invariant(!Environment.IsUnresolvableReference(this.realm, reference));
+      let referencedBase = reference.base;
+      let referencedName: string = (reference.referencedName: any);
+      if (typeof referencedName !== "string") {
+        throw new FatalError("TODO: do not know how to visit reference with symbol");
```

Comment:
Not your code, but since you are here, you never want a user visible error message to live inside a FatalError. Always call realm.handleError with a CompilerDiagnostic. In this case, however, I think that you are asserting that Environment.ResolveBinding is never going to return a Reference with a referencedName that is different from the name passed as its second parameter. That is best expressed as an invariant, along with a comment that clarifies the assumption and the reasoning for it.